File: golang/stack/calculator.go
package main

import "unicode"

func calculate(s string) int {
	stack := []int{}
	currentNumber := 0

	result := 0
	sign := 1

	for _, char := range s {
		if unicode.IsDigit(char) {
			currentNumber = currentNumber*10 + int(char-'0')
		} else if char == '+' {
			result += sign * currentNumber
			currentNumber = 0
			sign = 1
		} else if char == '-' {
			result += sign * currentNumber
			currentNumber = 0

			sign = -1
		} else if char == '(' {
			stack = append(stack, result)
			stack = append(stack, sign)

			result = 0
			sign = 1
		} else if char == ')' {

			result += sign * currentNumber
			currentNumber = 0
			sign = stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			prevResult := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			result = prevResult + sign*result
		}
	}

	result += sign * currentNumber

	return result
}

File: java/src/main/java/com/sanedge/interview/stack/Calculator.java
package com.sanedge.interview.stack;

import java.util.Stack;

public class Calculator {
  public int calculate(String s) {
    Stack<Integer> stack = new Stack<>();
    int currentNumber = 0;
    int result = 0;
    int sign = 1;

    for (int i = 0; i < s.length(); i++) {
      char ch = s.charAt(i);

      if (Character.isDigit(ch)) {
        currentNumber = currentNumber * 10 + (ch - '0');
      } else if (ch == '+') {
        result += sign + currentNumber;
        currentNumber = 0;
        sign = 1;
      } else if (ch == '-') {
        result += sign * currentNumber;
        currentNumber = 0;
        sign = -1;
      } else if (ch == '(') {
        stack.push(result);
        stack.push(sign);

        result = 0;
        sign = 1;
      } else if (ch == ')') {
        result += sign * currentNumber;
        currentNumber = 0;

        int poppedSign = stack.pop();
        int poppedResult = stack.pop();

        result = poppedResult + poppedSign * result;
      }
    }

    result += sign * currentNumber;

    return result;

  }

}
